### PR TITLE
Simplify function names in graph and flame views.

### LINKED ISSUE #377: Strip package names and args from Java method names to make graph and flamegraph more readable
The Java method names are currently rendered with their fully qualified names and parameters: for example, for [profile.pb.gz](https://github.com/google/pprof/files/1986407/profile.18.pb.gz) the graph view is

pprof graph

where method names are fully qualified, method arguments are displayed and use fully qualified type names, and dots in the fully qualified names are replaced with newlines. This produces display that is not very readable.

The flame graph display is better but far from perfect:

pprof flamegraph

It would be better if the package name and the arguments would be stripped from the display name and would only be displayed in the tooltip. Something like Stackdriver Profiler does:

7meawzmomxc

There was an attempt to do something like this in #367, see https://github.com/google/pprof/blob/520140b6bf47519c766e8380e5f094576347b016/internal/driver/flamegraph.go#L106, but it is only for flame graph and only for Go. We should come up with something that works for both graph and flame graph (not sure about other views) and for at least both Go and Java.

### PR DESCRIPTION
**Pull request by nolanmar511**
This fixes #377

Before this change:
image

image

After this change:
image

image

### CODE REVIEW

**Comment by aalexand** [comment]
@nolanmar511 please link to #377. Please note that that issue asks to still show the full name in the tooltip - will that be done?

File: internal/graph/graph.go [hunk 1]
```diff
@@ -420,6 +421,40 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
 	return selectNodesForGraph(nodes, o.DropNegative)
 }
 
+// SimplifyFunctionNames replaces function names in graph with shortened names.
```
Comment (aalexand) [inline internal/graph/graph.go]: No tests?..

Comment (nolanmar511) [reply internal/graph/graph.go]: Added test -- unfortunately no tests for the flame graph, though.

**Comment by nolanmar511** [comment]
Now only graph/flame are impacted, and the full name is shown in graph view and flame view tool tips.

**Comment by nolanmar511** [comment]
PTAL

File: internal/graph/graph.go [hunk 2]
```diff
@@ -420,6 +421,33 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
 	return selectNodesForGraph(nodes, o.DropNegative)
 }
 
+var javaFullRegExp = regexp.MustCompile(`^(?:[a-z]\w*\.)*[A-Z][\w\$]*\.(?:<init>|[a-z]\w*(?:\$\d+)?)(?:\(|$)`)
+var javaShortRegExp = regexp.MustCompile(`([A-Z][\w\$]*\.(?:<init>|[a-z]\w*(?:\$\d+)?))(?:|$)`)
+var goPrefixRegExp = regexp.MustCompile(`^(?:[a-z]\w*(\.\w*)*\/)+.`)
```
Comment (rauls5382) [inline internal/graph/graph.go]: Could these start with uppercase as well, or underscores?

File: internal/graph/graph.go [hunk 3]
```diff
@@ -420,6 +421,33 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
...
+var goPathRegExp = regexp.MustCompile(`^(?:[a-z]\w*(\.\w*)*\/)+`)
+var cppFullRegExp = regexp.MustCompile(`^(?:[_a-zA-Z]\w*\::)*_*[A-Z]\w*::~?[_a-zA-Z]\w*$`)
+var cppAnonymousPrefixRegExp = regexp.MustCompile(`^\(anonymous namespace\)::`)
+
+// GetNodeShortName returns a shortened version of a function's name.
+func GetNodeShortName(f string) string {
+	if goPrefixRegExp.MatchString(f) {
```
Comment (rauls5382) [inline internal/graph/graph.go]: How confident are we that we would have only a single match? Are there any names that could potentially match go, java and C++?

File: internal/graph/graph.go [hunk 4]
```diff
@@ -420,6 +421,33 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
...
 
+var javaFullRegExp = regexp.MustCompile(`^(?:[a-z]\w*\.)*[A-Z][\w\$]*\.(?:<init>|[a-z]\w*(?:\$\d+)?)(?:\(|$)`)
+var javaShortRegExp = regexp.MustCompile(`([A-Z][\w\$]*\.(?:<init>|[a-z]\w*(?:\$\d+)?))(?:|$)`)
+var goPrefixRegExp = regexp.MustCompile(`^(?:[a-z]\w*(\.\w*)*\/)+.`)
+var goPathRegExp = regexp.MustCompile(`^(?:[a-z]\w*(\.\w*)*\/)+`)
+var cppFullRegExp = regexp.MustCompile(`^(?:[_a-zA-Z]\w*\::)*_*[A-Z]\w*::~?[_a-zA-Z]\w*$`)
+var cppAnonymousPrefixRegExp = regexp.MustCompile(`^\(anonymous namespace\)::`)
+
+// GetNodeShortName returns a shortened version of a function's name.
+func GetNodeShortName(f string) string {
```
Comment (rauls5382) [inline internal/graph/graph.go]: Maybe ShortenFunctionName ? The comment describing this function talks about functions, not nodes.

Comment (nolanmar511) [reply internal/graph/graph.go]: Done.

File: internal/graph/graph.go (hunk 3 above)
Comment (nolanmar511) [reply internal/graph/graph.go]: I modified this function somewhat.

The go regexp (goPrefixRegExp), java regexp (javaFullRegExp), and cpp regexp () must include the beginning of the string, and java and cpp match to the end of the string. I think a function must have a "/" to match go, it must have a "::" to match c++ (and cannot have a "/"), and java can't have either a "/" or a "::"

So I'm fairly confident only one of these can ever be matched.

File: internal/driver/driver_test.go [hunk 5]
```diff
@@ -96,6 +96,8 @@ func TestParse(t *testing.T) {
 		{"peek=line.*01", "cpu"},
 		{"weblist=line[13],addresses,flat", "cpu"},
 		{"tags,tagfocus=400kb:", "heap_request"},
+		{"dot", "longFuncs"},
```
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/driver/driver_test.go]: Done.

File: internal/driver/driver_test.go [hunk 6]
```diff
@@ -96,6 +96,8 @@ func TestParse(t *testing.T) {
...
+		{"text", "longFuncs"},
```
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/driver/driver_test.go]: Done.

File: internal/driver/driver_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/driver/driver_test.go]: Done

File: internal/driver/driver_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/driver/driver_test.go]: Done.

File: internal/graph/graph_test.go [hunk 7]
```diff
@@ -398,3 +398,59 @@ func TestCreateNodes(t *testing.T) {
...
+		},
+		{
+			"(anonymous namespace)::Bar::Foo",
+			"Bar::Foo",
+		},
+	}
+	for _, tc := range testcases {
+		name := GetNodeShortName(tc.name)
+		if got, want := name, tc.want; got != want {
+			t.Errorf("for %s, got %q, want %q", tc.name, got, want)
```
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/graph/graph_test.go]: Done!
Moved the test, and I didn't fix it.

File: internal/graph/graph.go [hunk 8]
```diff
@@ -420,6 +421,33 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
...
+	if goPrefixRegExp.MatchString(f) {
+		match := goPathRegExp.ReplaceAllString(f, "")
+		return match
+	} else if javaFullRegExp.MatchString(f) {
+		match := javaShortRegExp.FindString(f)
+		if match != "" {
+			return match
+		}
+	} else {
+		name := cppAnonymousPrefixRegExp.ReplaceAllString(f, "")
```
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/graph/graph.go]: Done.

File: internal/graph/graph.go [hunk 9]
```diff
@@ -420,6 +421,33 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
 	return selectNodesForGraph(nodes, o.DropNegative)
 }
 
+var javaFullRegExp = regexp.MustCompile(`^(?:[a-z]\w*\.)*[A-Z][\w\$]*\.(?:<init>|[a-z]\w*(?:\$\d+)?)(?:\(|$)`)
```
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/graph/graph.go]: Done.

File: internal/graph/graph.go [hunk 10]
```diff
@@ -420,6 +421,33 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
...
+var javaShortRegExp = regexp.MustCompile(`([A-Z][\w\$]*\.(?:<init>|[a-z]\w*(?:\$\d+)?))(?:|$)`)
+var goPrefixRegExp = regexp.MustCompile(`^(?:[a-z]\w*(\.\w*)*\/)+.`)
+var goPathRegExp = regexp.MustCompile(`^(?:[a-z]\w*(\.\w*)*\/)+`)
+var cppFullRegExp = regexp.MustCompile(`^(?:[_a-zA-Z]\w*\::)*_*[A-Z]\w*::~?[_a-zA-Z]\w*$`)
+var cppAnonymousPrefixRegExp = regexp.MustCompile(`^\(anonymous namespace\)::`)
+
+// GetNodeShortName returns a shortened version of a function's name.
+func GetNodeShortName(f string) string {
+	if goPrefixRegExp.MatchString(f) {
+		match := goPathRegExp.ReplaceAllString(f, "")
```
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/graph/graph.go]: Done.

File: internal/graph/graph.go [hunk 11]
```diff
@@ -420,6 +421,33 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
...
+		return match
+	} else if javaFullRegExp.MatchString(f) {
```
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/graph/graph.go]: Done.

File: internal/graph/graph.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/graph/graph.go]: Tried something out for this. Let me know how to improve.

**Comment by nolanmar511** [comment]
PTAL

File: internal/graph/graph.go [hunk 12]
```diff
@@ -420,6 +421,35 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
 	return selectNodesForGraph(nodes, o.DropNegative)
 }
 
+var javaRegExp = regexp.MustCompile(`^(?:[a-z]\w*\.)*([A-Z][\w\$]*\.(?:<init>|[a-z]\w*(?:\$\d+)?))(?:(?:\()|$)`)
+var goRegExp = regexp.MustCompile(`^(?:[\w\-\.]+\/)+(.+)`)
+var cppAnonymousPrefixRegExp = regexp.MustCompile(`^(?:\(anonymous namespace\)::)(.*)`)
```
Comment (aalexand) [inline internal/graph/graph.go]: I wonder if this regular expression can be merged into cppRegExp, as an optional prefix. Then the loop below would simplify.

Comment (nolanmar511) [reply internal/graph/graph.go]: I tried merging it as an optional prefix, but then cases like "(anonymous namespace)::foo" are missed. 

I have a single working regex for now, I think.

File: internal/graph/graph.go [hunk 13]
```diff
@@ -420,6 +421,21 @@ func newTree(prof *profile.Profile, o *Options) (g *Graph) {
 	return selectNodesForGraph(nodes, o.DropNegative)
 }
 
+var javaRegExp = regexp.MustCompile(`^(?:[a-z]\w*\.)*([A-Z][\w\$]*\.(?:<init>|[a-z]\w*(?:\$\d+)?))(?:(?:\()|$)`)
```
[the remark this answers is not in the document]
Comment (nolanmar511) [reply internal/graph/graph.go]: Done. 
Also moved these to the top of the file.

**Pull request merged**